### PR TITLE
Introduce SPIFFE Bundle Map spec

### PR DESCRIPTION
**Pull request by evan2645**
This PR introduces the SPIFFE Bundle Map text that has been making its
way through SIG-Spec. A link to that document is below for posterity.

https://docs.google.com/document/d/13KHycYIfRC-g42aEfo4_4inF_WauCXOBgZAcUGuxdgs

### CODE REVIEW

**Comment by kfox1111** [comment]
LGTM! :)

File: standards/SPIFFE_Trust_Domain_and_Bundle.md [hunk 1]
```diff
@@ -197,6 +220,38 @@ Trust bundle #2 for example.com:
...
 In trust bundle #2, note that the `spiffe_sequence` parameter has been incremented and the second root certificate for `example.com` has been added. Once this new trust bundle is published and distributed, validators will accept SVIDs signed by either the original or the replacement root certificate. By publishing the replacement certificate well ahead of the expiration of the original certificate, validators have ample opportunity to refresh the trust bundle of example.com and learn of the pending replacement certificate.
 
+## Appendix B. SPIFFE Bundle Map Example
+The following is a sample SPIFFE Bundle Map:
+
+```
+{
+  “trust_domains”: {
+    "example.com": {
+      “sequence_number”: 12035488,
```
Comment (yt2985) [inline standards/SPIFFE_Trust_Domain_and_Bundle.md]: Should this be the `sequence_number` or the `spiffe_sequence`?

File: standards/SPIFFE_Trust_Domain_and_Bundle.md [hunk 2]
```diff
@@ -78,15 +81,32 @@ The `kty` parameter MUST be set, and its behavior follows [Section 4.1][6] of RF
...
+## 5. SPIFFE Bundle Map
+A SPIFFE Bundle Map is an object containing a collection of SPIFFE Bundles. The bundles it contains are considered authoritative for the corresponding trust domain name, which is also included.
+
+SPIFFE Bundle Maps are designed for use between SPIFFE control plane implementations, between SPIFFE implementations and the workloads they serve, as well as internally (i.e. as an implementation detail of a SPIFFE control plane).
+
+### 5.1. SPIFFE Bundle Map Format
+SPIFFE Bundle Maps are JSON-encoded structures, encapsulating one or more SPIFFE Bundles and some additional metadata. For an example of a complete SPIFFE Bundle Map, please see [Appendix B. SPIFFE Bundle Map Example](#appendix-b-spiffe-bundle-map-example).
+
+#### 5.1.1. Trust Domains
+A static “trust_domains” key MUST be set, and MAY be empty. Its contents are a map of SPIFFE Bundles, keyed by the trust domain name. For more information on what constitutes a valid trust domain name, please see [Section 2][1] of the SPIFFE ID specification.
```
Comment (arndt-s) [inline standards/SPIFFE_Trust_Domain_and_Bundle.md:93]: I'm thinking if something like "produced_at" with a timestamp value would be very helpful. If delivered asynchronously it'll be very hard to get a notion of freshness of it. A timestamp would go a long way for debugging purposes.

Alternatively, explicitly allowing additional keys to be present could make something like it an implementers choice.

File: standards/SPIFFE_Trust_Domain_and_Bundle.md [hunk 3]
```diff
@@ -95,6 +115,9 @@ Continuing the above example where a naïve implementation imports a particular
...
 
+### 6.3. JSON Key Duplication
+[RFC 8259][12], which serves as the standard for JSON data format, states that the names within an object SHOULD be unique. Furthermore, it emphasizes that when the names within a JSON object are not unique, the behavior of software that receives such an object is unpredictable. This behavior poses an unacceptable security risk when the JSON format is used to represent SPIFFE Bundle Maps, as the presence of duplicate trust domain names could lead to the incorrect bundle being selected for use in validating SVIDs. As such, it is important that libraries parsing SPIFFE bundle maps employ duplicate key detection.
+
```
Comment (PieterKas) [inline standards/SPIFFE_Trust_Domain_and_Bundle.md:120]: SPIFFE bundles are JWK sets and RFC7515 which gives guidance on key provenance and trust ( https://datatracker.ietf.org/doc/html/rfc7517#section-9.1). 

Should there be similar considerations for SPIFFE bundle maps to regarding their origin and integrity?

For example, are there any risks that may result from an adversary adding or removing SPIFFE bundles to a Bundle Map? Are there precautions the recipient should take?

**Comment by cesarghali** [comment]
Hi @evan2645, I'm curious of there are any updates about the status of this PR. Also, when do you think it can be merged? Thank you!

**Comment by evan2645** [comment]
Hi @cesarghali!

We have a SIG-Spec call scheduled tomorrow related to this PR and some other work. Anybody who intends to use this spec should join as the primary hesitance thus far is around lack of community feedback. Please see [SIG-Spec](https://github.com/spiffe/spiffe/blob/main/community/sig-spec/README.md) calendar for call information. More than happy to hold follow-on calls as well if the time doesn't work well for you.

**Comment by cesarghali** [comment]
Thank you @evan2645 for the information. I will attend the SIG meeting tomorrow!

File: standards/SPIFFE_Trust_Domain_and_Bundle.md (hunk 2 above)
Comment (yamandabbagh) [reply standards/SPIFFE_Trust_Domain_and_Bundle.md:93]: Thank you @arndt-s, I agree having a timestamp has some value. We circulated this idea of having a timestamp or some sort of a refresh hint to the map during the Spiffe community meetings. The conclusion we reached that while adding something to show freshness has some value, keeping the Spec simple and leaving this detail up to the implementers would keep the specification simple and hence increase adoption. 

If we add to the spec a required timestamp/freshness field we would need an agreement about how this field would be consumed, when maps expire, how often we need to refresh. And just adding an optional field without further details and why it is important would add complexity to the Spec with minimal benefit. What do you think?

File: standards/SPIFFE_Trust_Domain_and_Bundle.md (hunk 3 above)
Comment (yamandabbagh) [reply standards/SPIFFE_Trust_Domain_and_Bundle.md:120]: That's an important point. The same trust considerations that apply to a single SPIFFE bundle, where an adversary could manipulate the contents, should also apply to the bundle map.  Both producers and consumers of these bundles must implement measures to ensure the integrity of the bundle map. The specifics of these measures will vary depending on the implementation.  To address this, we could add a security consideration to the specification emphasizing the importance of bundle map integrity, while leaving the implementation details to developers. What are your thoughts on this approach?

**Comment by cesarghali** [comment]
Overall LGTM! Thanks everyone for working on this PR. Merging it would unblock planned implementations and avoid short-term and long-term divergence from the specs.

File: standards/SPIFFE_Trust_Domain_and_Bundle.md (hunk 2 above)
Comment (arndt-s) [reply standards/SPIFFE_Trust_Domain_and_Bundle.md:93]: Agreed on the focus of keeping it simple to easy use and adoption.

Closing this comment. 
Update: looks like I can't close it. Please consider closed.

File: standards/SPIFFE_Trust_Domain_and_Bundle.md (hunk 1 above)
Comment (evan2645) [reply standards/SPIFFE_Trust_Domain_and_Bundle.md]: It should be `spiffe_sequence`

File: standards/SPIFFE_Trust_Domain_and_Bundle.md (hunk 3 above)
Comment (evan2645) [reply standards/SPIFFE_Trust_Domain_and_Bundle.md:120]: Thank you for the observation @PieterKas .. I feel it's an important one. We've discussed this in the past when designing SPIFFE Bundle spec , but I think now things have changed with the introduction of this spec and other patterns. We probably need to have a closer look at the bundle spec as well. Could you please open a dedicated issue so we can discuss further and more broadly?

Comment (PieterKas) [reply standards/SPIFFE_Trust_Domain_and_Bundle.md:120]: Sounds good to me.

**Comment by evan2645** [comment]
Thanks to all who contributed to the development of this spec. I think it is ready to go. @azdagron please take a look

**Review by azdagron — APPROVED** [review]
The proposed addition looks good!

**Pull request merged**